comp/core/gui/guiimpl/gui.go

```diff
@@ -228,9 +234,45 @@ func renderIndexPage(w http.ResponseWriter, _ *http.Request) {
 	}
 }
 
+func serveAsset(w http.ResponseWriter, req *http.Request, assetPath string) ([]byte, error) {
+	var rdr io.ReadCloser
+	isGzipped := true
+	rdr, err := viewsFS.Open(assetPath + ".gz")
+	if err != nil && os.IsNotExist(err) {
+		isGzipped = false
+		rdr, err = viewsFS.Open(assetPath)
+	}
+	if err != nil {
+		return nil, err
+	}
+	defer rdr.Close()
+
+	if isGzipped {
+		// if client supports gzip, return as-is
+		if strings.Contains(req.Header.Get("Accept-Encoding"), "gzip") {
+			w.Header().Set("Content-Encoding", "gzip")
+		} else {
+			// ungzip on the fly
+			gzReader, err := gzip.NewReader(rdr)
+			if err != nil {
+				return nil, err
+			}
+			// ensure underlying io.Reader is closed
+			defer gzReader.Close()
+			rdr = gzReader
+		}
+	}
+
+	data, err := io.ReadAll(rdr)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
 func serveAssets(w http.ResponseWriter, req *http.Request) {
-	path := path.Join("views", "private", req.URL.Path)
-	data, err := viewsFS.ReadFile(path)
+	assetPath := path.Join("views", "private", req.URL.Path)
+	data, err := serveAsset(w, req, assetPath)
```

Comment:
What if we group all header logic together? 

Here is a suggestion of how it could look like:

```go
func serveAsset(gzipSupport bool, assetPath string) ([]byte, error) {
	var rdr io.ReadCloser
	rdr, err := viewsFS.Open(assetPath + ".gz")
	if err != nil && os.IsNotExist(err) {
		rdr, err = viewsFS.Open(assetPath)
	}
	if err != nil {
		return nil, err
	}
	defer rdr.Close()

	if !gzipSupport {
		// ungzip on the fly
		gzReader, err := gzip.NewReader(rdr)
		if err != nil {
			return nil, err
		}
		// ensure underlying io.Reader is closed
		defer gzReader.Close()
		rdr = gzReader
	}

	data, err := io.ReadAll(rdr)
	if err != nil {
		return nil, err
	}
	return data, nil
}

func serveAssets(w http.ResponseWriter, req *http.Request) {
	assetPath := path.Join("views", "private", req.URL.Path)
	gzipSupport := strings.Contains(req.Header.Get("Accept-Encoding"), "gzip")
	data, err := serveAsset(gzipSupport, assetPath)
	if err != nil {
		if os.IsNotExist(err) {
			http.Error(w, err.Error(), http.StatusNotFound)
		} else {
			http.Error(w, err.Error(), http.StatusInternalServerError)
		}
		return
	}
	
	if gzipSupport {
		w.Header().Set("Content-Encoding", "gzip")
	}
	
	w.Header().Set("Content-Type", mime.TypeByExtension(filepath.Ext(assetPath)))
	w.Header().Set("Content-Length", strconv.Itoa(len(data)))
	_, _ = w.Write(data)
}

```
 Let me know what you think :smi

Reply:
Updated